Sources/Shared/LocalAuthentication/BiometricsAuthenticationService.swift

```diff
@@ -0,0 +1,109 @@
+
+    /// Adds the protection overlay and tries to authenticate
+    func checkForBiometrics(controller: UIViewController) {
+        // Means biometrics lock is on top
+        if let previousRootViewController {
+            authenticate()
+            return
+        }
+
+        guard context == nil else { return }
+        protectAppIfNeeded(controller: controller) { [weak self] in
+            if Current.settingsStore.biometricsRequired {
+                self?.authenticate()
+            }
+        }
+    }
+
+    /// Adds the protection overlay
+    func protectAppIfNeeded(controller: UIViewController, completion: (() -> Void)?) {
+        if Current.settingsStore.biometricsRequired {
+            addBiometricOverlayProtection(controller: controller) {
+                completion?()
+            }
+        } else {
+            removeProtectionOverlay()
+        }
+    }
+
+    private func authenticate() {
+        var error: NSError?
+        context?.invalidate()
+        context = LAContext()
+        if context?.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) ?? false {
+            authenticate(policy: .deviceOwnerAuthentication)
+        } else {
+            Current.Log
+                .error(
+                    "Failed to authenticate with biometrics, user context can't evaluate policy for device ownership"
+                )
+        }
+    }
+
+    private func removeProtectionOverlay() {
+        guard let window, let previousRootViewController else { return }
+        DispatchQueue.main.async { [weak self] in
+            window.rootViewController = previousRootViewController
+            self?.previousRootViewController = nil
+        }
+    }
+
+    private func authenticate(policy: LAPolicy) {
+        context?
+            .evaluatePolicy(
+                policy,
+                localizedReason: L10n.SettingsDetails.General.Security
+                    .authenticationPolicy
+            ) { [weak self] authorized, error in
+                self?.didFinishAuthentication(authorized: authorized)
+
+                if let error {
+                    Current.Log.error(["Failed to evaluate policy for authentication": error.localizedDescription])
+                }
+            }
+    }
+
+    private func didFinishAuthentication(authorized: Bool) {
+        if authorized {
+            context?.invalidate()
+            context = nil
+            removeProtectionOverlay()
+        }
+    }
+
+    private func addBiometricOverlayProtection(controller: UIViewController, completion: @escaping () -> Void) {
+        guard #available(iOS 13.0, *) else { return }
+
+        let overlayController = UIHostingController(rootView: BiometricsView.build(delegate: self))
+        overlayController.modalPresentationStyle = .overCurrentContext
+
+        guard let window = controller.view.window,
+              previousRootViewController == nil else { return }
+
+        self.window = window
```

Comment:
what happens if there are multiple windows open? e.g. multi-scene support on iPad.

i think you probably want to either collate all the windows during the lock phase, or make this chain off the WebViewWindowController itself, but synchronize the unlock status

Reply:
I think I will start over with this implementation, I liked your idea of sync status. Another question... should we allow shortcut actions if the app is locked?

Reply:
I think it becomes really hard to do locking if we disallow extensions to do work. Most of them do not support being able to prompt the user for permission.

We could either:
- tell the user it won't lock extensions
- disallow extensions entirely if lock is enabled (where we cannot prompt)
- allow the user to choose which extensions should be allowed

Maybe we could add more control in app but I think the first one sounds easiest for now.